gax/retry_policy.h

```diff
@@ -0,0 +1,119 @@
+// Copyright 2019 Google Inc.  All rights reserved
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_API_CODEGEN_GAPIC_RETRY_POLICY_H_
+#define GOOGLE_API_CODEGEN_GAPIC_RETRY_POLICY_H_
+
+#include <chrono>
+#include <memory>
+
+namespace google {
+namespace gapic {
+namespace gax {
+
+/**
+ * Define the interface for controlling how clients retry RPC operations.
+ *
+ * Idempotent (and certain non-idempotent) operations can be retried
+ * transparently to the user. However, we need to give the users enough flexiblity
+ * to control when to stop retrying.
+ *
+ * The application provides an instance of this class when the client is created.
+ */
+template <typename StatusType, typename RetryablePolicy>
+class RetryPolicy {
```

Comment:
What about the `Setup` method?

Reply:
Setup doesn't currently have any testable semantics; it's heavily reliant on what gax::Context winds up being and on other aspects of the API that have yet to be resolved, e.g. owning thread for asynchronous calls. I can add a stub so that we don't forget about it, but I'm hesitant to for TDD purity reasons.

Reply:
SG, let's omit for now.